internal/bindings/server.go

```diff
@@ -275,11 +276,15 @@ type fileConn interface {
 func connectWithDial(handle C.uintptr_t, address *C.char, fd *C.int) C.int {
 	connectLock.Lock()
 	defer connectLock.Unlock()
+
 	dial := connectRegistry[handle]
+	ctx := contextRegistry[handle]
+
 	// TODO: make timeout customizable.
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	dialCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
```

Comment:
As this function extends the node's ctx with a 5s timeout context, presumably, for each dial attempt, then we can say that the node's ctx is a long-lived one. But, now that we are using the app's context to cancel the node's proxy go routine, can/should we also use the app's context to cancel (prematurely) the node's dial attempt also (and avoid the need for a per-node ctx).

Basically what I'm aiming for is consistency. If we have a long-live node context, then it should be used as a basis for all node related operations (in this case the dial attempt and the long-lived proxy routine).

But if we are now using the app's context for the long-lived node proxy routine, then for consistency/ease of understanding, we should also use the app's context for the basis of the 5s dial timeout context too.

The question is, I think, is a node stopped (which can cancel its node ctx) when the app continues running, and in those situations, should we have cancelled any ongoing dial operations as well as any long-live proxy routines?

Reply:
I left the contexts separate because we can create, start up and close a dqlite node without using the app package at all, like we do in LXD. Though in a way, LXD is taking the role of the app package in that case, so perhaps we can add a `WithContext` option to the options passed to `dqlite.New` that overrides the background context as the parent context for the dqlite node. Then if it makes sense for LXD, we can use `WithContext` to set the gateway context to the parent context of the dqlite node.

Of course, we can force consideration of the context of dqlite by just adding a context argument to `dqlite.New`, but this would mean we would have to update LXD to use this before we can update the dqlite version.

Reply:
I'm fine with just having a node context, but we should then ensure the proxy() function is passed it rather than the app context as it does now.

Reply:
Since the dial functions were being configured before the dqlite server was starting, I moved the logic for configuring the dial funcs and starting up the proxy into the dqlite package itself. So now the context passed to the dial function (the one used for the proxies) has the node context as its parent. 

I moved the node.go file into its own directory just to make things a bit cleaner (as now there's 4 files in there), but this changes the path that dqlite is currently imported.

Reply:
Sorry, I didn't realize the diff would be so horrible. I've moved the dqlite package stuff back out into the root of the project.

Bit confused about your suggestion for `SetDialFunc`. There's unfortunately 2 `Node` types which makes things a bit confusing.  I think you mean adding context fields to the `Node` struct for the bindings, as that's the owner of `SetDialFunc`. This works, but I think the `contextRegistry` would still be necessary since `connectWithDial` will eventually need to call the dial function with the right context and it's not a method on `Node`.

Reply:
What do you think about creating the context in `Start` instead? I'm not so sure about starting a 5 second context in `New`, which will be counting down right away, even though we're not even dialing yet.

Reply:
We could have it in Start too I guess, but New seems better as it will be initialised even if Start isn't called.

Reply:
So now,`dqlite.Node` creates a new context from `context.Background()` in `New`, and passes it to `server.NewNode`, which stores it in the context registry.

Reply:
I think I may be confusing the situation, when I get a chance I'll take a crack at this and hopefully will be able to provide better explanations :)